contributor-docs/adrs/adr-013-primer-experimental-location.md

```diff
@@ -0,0 +1,62 @@
+# ADR 013: Location of experimental react components maintained by Primer
+
+## Status
+
+❌ Not Adopted, `<issue with implementation details goes here>`
+
+&nbsp;
+
+## Context: Experimental components (or drafts) in primer/react
+
+As we work on new components (or rewrites of old components), we often need to start with a "experiment" and not include it in the semantically versioned (semver-ed) main bundle.
+
+[ADR-006](./adr-006-drafts.md) explains our use of "drafts", here are the relevant parts:
+
+- When we are not sure about the API of a component, we put it in a different bundle: "primer/react/drafts". [ADR 010: Merge drafts status into experimental](./adr-010-drafts-are-experimental) recommends changing the name to `experimental`
+- The contract with consumers is that you are opting into a component in which we don't have 100% confidence yet, the API and structure can change.
+
+In [ADR 008: Lowering the barrier to entry](./adr-008-experimental-components) (Sep, 2022), we decided to create room to share experimental components between memex and the monolith with a new repository ([github/experimental-react-components](https://github.com/github/experimental-react-components)). Now that [memex is part of the monolith](https://github.com/github/planning-tracking/issues/1755), that [repository has been archived](https://github.com/github/platform-ux/issues/1248) (Feb, 2023).
+
+Which leaves us with: Where should experimental components maintained by Primer live?
+
+&nbsp;
+
+### Problems with publishing experimental components in @primer/react npm package:
+
+**Incompatible with semver**: Because we haven't reached the final API or structure, we need to ship breaking changes to experimental components. In the current setup, these breaking changes are part of minor or patch releases to primer/react.
+
+On the consumer side, this makes every upgrade a potentially breaking upgrade nullifying the benefits of semver.
+
+In the case of multiple teams working on the monolith, this can mean finding breaking changes for a new experimental component in a part of the app that you're not familiar with.
+
+&nbsp;
+
+## Decision
+
+1. Experimental components, as defined by the [component lifecycle](https://primer.style/contribute/component-lifecycle#experimental), should not be part of semantically versioned npm package for `@primer/react`.
+
+2. Each experimental component should be independently published and versioned with semver. Breaking changes should be tagged as major versions. Example: `@primer/react-experimental-underlinenav`.
+
+3. The source code for all experimental components should continue to live in the same repository ([primer/react](https://github.com/primer/react)).
+
+### Tradeoffs
+
+1. The responsibility of using compatible versions of experimental components and primer/react is shifted to the consumer.
```

Comment:
_Kind of_. I think this is an acceptable tradeoff, but I'm curious: what happens when they reach stable?

I would love to see Primer move to a monorepo that houses all stable _and_ experimental components as individual packages, with the `@primer/react` package essentially importing and re-exporting all components. That gives us a few things:

1. It would let experimental components continue to be separate packages once they become stable, thereby eliminating the need to update usages in code (i.e. change imports from the individual package to the bundled package)
2. It would allow for all the code to live in one place, making it easier to share and re-use components
3. It would allow for consumers to install/import _only what they need_. While tree-shaking covers us today, this is still a nice to have to be able to have dependencies on a smaller subset of components, in the event of incompatibility, breaking change churn, and bugs that might pop up.

Reply:
I like the idea in theory, but I really don't like the maintenance overhead it comes with of matching compatible versions both in the library and application.

wrote an unpublished rant about this earlier: https://www.sid.st/unpolished/component-versioning

Reply:
That write up is a good summary, but I think it also misses a key benefit: If I am only consuming `Button`, I don't care about upgrades to other components, nor do I want to have to deal with the potential of some side effect causing `Button` to behave in an unexpected way. Publishing individual packages solves that.

Also, wrt Composite components (e.g. your Button/Dialog example), versioning is a bit more nuanced. If the Dialog implements the new button, _does_ it require a major version bump? I would argue no, unless the API changes. If the implementation details are hidden away from the user, a minor or patch version could be used.

However on the flip side, there is still the larger dependency management question/mess. If I upgrade `Dialog`, but want to keep my old usage of `Button`, now I have to manage two versions of the component. Not _that_ bad, but would be compounded by the addition of more components.